File: src/lib/puzzle3d/candidates3d.ts
import { getIndex3D } from './validator3d';

/** Get candidate values (1–size) for a cell, excluding row/col/box within layer AND pillar */
export function getCandidates3D(grid: (number | null)[], flatIndex: number, size: number): Set<number> {
  const layerSize = size * size;
  const layer = Math.floor(flatIndex / layerSize);
  const layerIndex = flatIndex % layerSize;
  const row = Math.floor(layerIndex / size);
  const col = layerIndex % size;
  const boxSize = Math.sqrt(size);
  const depth = size; // NxNxN: depth === size

  const used = new Set<number>();

  // Row in layer
  for (let c = 0; c < size; c++) {
    const v = grid[getIndex3D(layer, row, c, size)];
    if (v !== null) used.add(v);
  }

  // Col in layer
  for (let r = 0; r < size; r++) {
    const v = grid[getIndex3D(layer, r, col, size)];
    if (v !== null) used.add(v);
  }

  // Box in layer
  const boxRow = Math.floor(row / boxSize) * boxSize;
  const boxCol = Math.floor(col / boxSize) * boxSize;
  for (let r = boxRow; r < boxRow + boxSize; r++) {
    for (let c = boxCol; c < boxCol + boxSize; c++) {
      const v = grid[getIndex3D(layer, r, c, size)];
      if (v !== null) used.add(v);
    }
  }

  // Pillar (same row/col across all layers)
  for (let d = 0; d < depth; d++) {
    const v = grid[getIndex3D(d, row, col, size)];
    if (v !== null) used.add(v);
  }

  const candidates = new Set<number>();
  for (let v = 1; v <= size; v++) {
    if (!used.has(v)) candidates.add(v);
  }
  return candidates;
}

File: love/src/puzzle/candidates3d.lua
-- Candidate computation for 3D sudoku.
-- A cell's candidates exclude: row, col, box within its layer, AND its pillar.

local V3 = require "src.puzzle.validator3d"

local C3 = {}

-- Returns { [v] = true } of valid candidates for cells[flat1].
function C3.get(cells, flat1, n)
  if cells[flat1] ~= nil then return {} end

  local layer, row, col = V3.decompose(flat1, n)
  local bs    = math.floor(math.sqrt(n) + 0.5)
  local depth = n

  local used = {}

  for c = 0, n - 1 do
    local v = cells[V3.compose(layer, row, c, n)]
    if v then used[v] = true end
  end
  for r = 0, n - 1 do
    local v = cells[V3.compose(layer, r, col, n)]
    if v then used[v] = true end
  end

  local br = math.floor(row / bs) * bs
  local bc = math.floor(col / bs) * bs
  for r = br, br + bs - 1 do
    for c = bc, bc + bs - 1 do
      local v = cells[V3.compose(layer, r, c, n)]
      if v then used[v] = true end
    end
  end

  for d = 0, depth - 1 do
    local v = cells[V3.compose(d, row, col, n)]
    if v then used[v] = true end
  end

  local candidates = {}
  for v = 1, n do
    if not used[v] then candidates[v] = true end
  end
  return candidates
end

-- Returns sorted candidate list for display.
function C3.get_list(cells, flat1, n)
  local set  = C3.get(cells, flat1, n)
  local list = {}
  for v = 1, n do if set[v] then list[#list + 1] = v end end
  return list
end

return C3
